ethereumj-core/src/main/java/org/ethereum/vm/DataWord.java

```diff
@@ -22,77 +22,108 @@
+     */
+    private DataWord(byte[] data) {
+        if (data == null || data.length != 32) throw new RuntimeException("Input byte array should have 32 bytes in it!");
+        this.data = data;
     }
 
-    public DataWord(int num) {
-        this(ByteBuffer.allocate(4).putInt(num));
-    }
+    public static DataWord of(byte[] data) {
+        if (data == null || data.length == 0) {
+            return DataWord.ZERO;
+        }
 
-    public DataWord(long num) {
-        this(ByteBuffer.allocate(8).putLong(num));
+        boolean allExceptLastIsZero = true;
+        for (int i = 0; i < (data.length - 1); ++i) {
+            if (data[i] != 0) {
+                allExceptLastIsZero = false;
+                break;
+            }
+        }
+        if (allExceptLastIsZero) {
+            if (data[data.length - 1] == 0) return DataWord.ZERO;
+            if (data[data.length - 1] == 1) return DataWord.ONE;
+        }
+
+        if (data.length == 32)
+            return new DataWord(Arrays.copyOf(data, data.length));
+        else if (data.length <= 32) {
+            byte[] bytes = new byte[32];
+            System.arraycopy(data, 0, bytes, 32 - data.length, data.length);
+            return new DataWord(bytes);
+        } else {
+            throw new RuntimeException(String.format("Data word can't exceed 32 bytes: 0x%s", ByteUtil.toHexString(data)));
+        }
     }
 
-    private DataWord(ByteBuffer buffer) {
-        final ByteBuffer data = ByteBuffer.allocate(32);
-        final byte[] array = buffer.array();
-        System.arraycopy(array, 0, data.array(), 32 - array.length, array.length);
-        this.data = data.array();
+    public static DataWord of(ByteArrayWrapper wrappedData) {
+        return of(wrappedData.getData());
     }
 
     @JsonCreator
-    public DataWord(String data) {
-        this(Hex.decode(data));
+    public static DataWord of(String data) {
+        return of(Hex.decode(data));
     }
 
-    public DataWord(ByteArrayWrapper wrappedData){
-        this(wrappedData.getData());
+    public static DataWord of(byte num) {
+        byte[] bb = new byte[32];
+        bb[31] = num;
+        return new DataWord(bb);
+
     }
 
-    public DataWord(byte[] data) {
-        if (data == null)
-            this.data = ByteUtil.EMPTY_BYTE_ARRAY;
-        else if (data.length == 32)
-            this.data = data;
-        else if (data.length <= 32)
-            System.arraycopy(data, 0, this.data, 32 - data.length, data.length);
-        else
-            throw new RuntimeException("Data word can't exceed 32 bytes: " + data);
+    public static DataWord of(int num) {
+        return of(ByteBuffer.allocate(4).putInt(num).array());
+    }
+
+    public static DataWord of(long num) {
+        return of(ByteBuffer.allocate(8).putLong(num).array());
     }
 
     public byte[] getData() {
-        return data;
+        return Arrays.copyOf(data, data.length);
```

Comment:
I'm still suggesting returning internal array here (mentioning this in method docs). It's a normal practice (e.g. take a look at `java.nio.ByteBuffer.array()`